### PR TITLE
train code for noise suppression model 

### PR DESCRIPTION
**Pull request by krodyush**

### CODE REVIEW

**Comment by Ilya-Krylov** [comment]
run tests

**Comment by Ilya-Krylov** [comment]
run tests

**Comment by morkovka1337** [comment]
Please, fix linter issues, provide tests and `init_venv.sh` for correct CI testing. As part of it, you will need a small toy dataset and the `setup.py` file. Also, I can not find a dedicated export code.

**Comment by krodyush** [comment]
can you point to the example?

**Comment by morkovka1337** [comment]
You can look at [face antispoofing](https://github.com/openvinotoolkit/training_extensions/tree/develop/misc/pytorch_toolkit/face_antispoofing) and [text recognition](https://github.com/openvinotoolkit/training_extensions/tree/develop/misc/pytorch_toolkit/text_recognition) examples.

**Comment by Ilya-Krylov** [comment]
run tests

**Comment by aiarkinx** [comment]
run tests

**Comment by aiarkinx** [comment]
run tests

**Comment by aiarkinx** [comment]
run tests

**Comment by aiarkinx** [comment]
run tests

**Comment by morkovka1337** [comment]
Please, fix tests and linter issues.

**Comment by krodyush** [comment]
run tests

File: misc/pytorch_toolkit/noise_suppression/README.md [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+--total_train_batch_size=128 \
+--dns_datasets=<dns-challenge-dir>/datasets \
+--eval_data=<dns-challenge-dir>/datasets/ICASSP_dev_test_set/track_1/synthetic
+```
+
+As a result the trained model has to be located in newly created folder 'result_model'.
+
+## Convert a Models to OpenVINO™ format for Demo
+
+The script together with pytorch model also save ONNX\* model into the same output folder. This ONNX\* model can be converted to OpenVINO™ format
```
Comment (morkovka1337) [inline misc/pytorch_toolkit/noise_suppression/README.md:55]:
```suggestion
The script also saves the ONNX\* model into the same output folder together with the PyTorch\* model . This ONNX\* model can be converted to OpenVINO™ format
```

File: misc/pytorch_toolkit/noise_suppression/README.md [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+```bash
+mo.py --input_model result_model/model.onnx
+```
+
+After conversion them to the OpenVINO™ format you can try model using
```
Comment (morkovka1337) [inline misc/pytorch_toolkit/noise_suppression/README.md:61]:
```suggestion
After conversion to the OpenVINO™ format you can try model using
```

File: misc/pytorch_toolkit/noise_suppression/README.md [hunk 3]
```diff
@@ -0,0 +1,63 @@
+# Noise Suppression
+
+This is implementation of the train script in PyTorch\*.
```
Comment (morkovka1337) [inline misc/pytorch_toolkit/noise_suppression/README.md:3]:
```suggestion
This is an implementation of the train script in PyTorch\*.
```

File: misc/pytorch_toolkit/noise_suppression/README.md [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+The script allows to train noise suppression model on [DNS-Challenge dataset](https://github.com/microsoft/DNS-Challenge).
```
Comment (morkovka1337) [inline misc/pytorch_toolkit/noise_suppression/README.md:4]:
```suggestion
The script allows to train a noise suppression model on the [DNS-Challenge dataset](https://github.com/microsoft/DNS-Challenge).
```

File: misc/pytorch_toolkit/noise_suppression/README.md [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+1. [Requirements](#requirements)
+2. [Preparation](#preparation)
+3. [Train model](#train-model)
+4. [Convert a Model to OpenVINO™ format for Demo](#convert-a-models-to-openvino-format-for-demo)
+
+
+## Requirements
+
+To run scrips the Python\* 3.6 has to be installed on machine.
```
Comment (morkovka1337) [inline misc/pytorch_toolkit/noise_suppression/README.md:18]:
```suggestion
To run the script Python\* 3.6 has to be installed on the machine.
```

File: misc/pytorch_toolkit/noise_suppression/README.md [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+To install the required packages, run the following:
+
+```bash
+pip install -r requirements.txt
+```
+
+## Preparation
+
+The [DNS-Challenge dataset](https://github.com/microsoft/DNS-Challenge) is used to train the model.
+You can get the data by clone repository
```
Comment (morkovka1337) [inline misc/pytorch_toolkit/noise_suppression/README.md:28]:
```suggestion
You can get the data by cloning the repository
```

File: misc/pytorch_toolkit/noise_suppression/README.md [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+```git clone https://github.com/microsoft/DNS-Challenge.git <dns-challenge-dir>```
+
+The train script also allows to evaluate model on synthetic ICASSP_dev_test_set. The 'icassp2021-final' branch of DNS-Challenge repo has to be checkout to enable this option
```
Comment (morkovka1337) [inline misc/pytorch_toolkit/noise_suppression/README.md:31]:
```suggestion
The train script also allows to evaluate model on synthetic ICASSP_dev_test_set. The 'icassp2021-final' branch of the DNS-Challenge repo has to be checked-out to enable this option
```

File: misc/pytorch_toolkit/noise_suppression/README.md [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+
+
+## Train Model
+
+After you get the data, you can train the model.
+To do this you may use the followed command line:
```
Comment (morkovka1337) [inline misc/pytorch_toolkit/noise_suppression/README.md:37]:
```suggestion
To do this you may use the following command line:
```

File: misc/pytorch_toolkit/noise_suppression/metrics.py [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+        else:
+            l = 10 * torch.log10(target_pow + reg_pow) - 10 * torch.log10(noise_pow + reg_pow)
+        if inv_flag:
+            l = -l
+    else:
+        if inv_flag:
+            l = (noise_pow + reg_pow) / (target_pow + reg_pow)
+        else:
+            l = (target_pow + reg_pow) / (noise_pow + reg_pow)
+    return l
```
Comment (morkovka1337) [inline misc/pytorch_toolkit/noise_suppression/metrics.py:53]:
```suggestion
    return l

```
Newline in the end

File: misc/pytorch_toolkit/noise_suppression/model_base.py [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+        return model
+
+    @classmethod
+    def from_pretrained(cls, dir_name):
+        weights_name = os.path.join(dir_name, cls.WEIGHTS_NAME)
+        config_name = os.path.join(dir_name, cls.CONFIG_NAME)
+        model = cls.from_config(config_name)
+        state_dict = torch.load(weights_name, map_location="cpu")
+        model.load_state_dict(state_dict)
+        return model
```
Comment (morkovka1337) [inline misc/pytorch_toolkit/noise_suppression/model_base.py:64]:
```suggestion
        return model

```
newline in the end of the file

File: misc/pytorch_toolkit/noise_suppression/models.py [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+from model_base import BaseDNSModel
+from model_poconetlike import PoCoNetLikeModel
+
+logging.basicConfig(format='%(asctime)s %(levelname)s %(name)s %(message)s',
+                    datefmt='%Y-%m-%d %H:%M:%S',level=logging.INFO)
+logger = logging.getLogger('{} models'.format(os.getpid()))
+def printlog(*args):
+    logger.info(' '.join([str(v) for v in args]))
+
+#MODELS = {cls.__name__:cls for cls in globals().values() if isinstance(cls, type) and issubclass(cls, BaseDNSModel)}
```
Comment (morkovka1337) [inline misc/pytorch_toolkit/noise_suppression/models.py:13]:
```suggestion

```

File: misc/pytorch_toolkit/noise_suppression/states.py [hunk 12]
```diff
@@ -0,0 +1,36 @@
+import torch
+
+def get_shape(x):
+    return list(x.shape)
+    #this is to get more compact and human readable but unresizable onnx model
+    #return [int(s) for s in x.shape]
```
Comment (morkovka1337) [inline misc/pytorch_toolkit/noise_suppression/states.py:6]: This could be imported from the `model.py` or moved to some common file like `utils.py`

File: misc/pytorch_toolkit/noise_suppression/states.py [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+
+        #add left part of x
+        x_padded = torch.cat([x_pad, x], dim)
+
+        #get right part for padding on next iter
+        x_splited = torch.split(x_padded, [get_shape(x_padded)[dim]-size, size], dim=dim)
+
+        self.update(x_splited[-1:])
+
+        return x_splited[0] if shift_right else x_padded
```
Comment (morkovka1337) [inline misc/pytorch_toolkit/noise_suppression/states.py:36]:
```suggestion
        return x_splited[0] if shift_right else x_padded

```
newline

File: misc/pytorch_toolkit/noise_suppression/utils.py [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+    torch.distributed.broadcast(data_tensor, rank_src)
+
+    if rank != rank_src:
+        printlog(
+            "Receive data and unpack {} bytes by {} process".format(size_tensor.item(), rank))
+        data_buf = data_tensor.cpu().numpy().tobytes()
+        data = pickle.loads(data_buf)
+    torch.distributed.barrier()
+
+    return data
```
Comment (morkovka1337) [inline misc/pytorch_toolkit/noise_suppression/utils.py:100]:
```suggestion
    return data

```
newline

**Review by morkovka1337 — CHANGES_REQUESTED** [review]
Please, also fix year in the license headers and add headers where they are missing.

File: misc/pytorch_toolkit/noise_suppression/states.py (hunk 12 above)
Comment (krodyush) [reply misc/pytorch_toolkit/noise_suppression/states.py]: moved to utils

**Comment by krodyush** [comment]
run tests

**Comment by krodyush** [comment]
run tests

**Comment by krodyush** [comment]
run tests

**Comment by krodyush** [comment]
run tests

**Comment by krodyush** [comment]
run tests

**Comment by krodyush** [comment]
run tests

**Comment by krodyush** [comment]
run tests

**Comment by krodyush** [comment]
run tests

File: misc/pytorch_toolkit/noise_suppression/train.py [hunk 15]
```diff
@@ -0,0 +1,572 @@
...
+
+        self.train_batch_size = self.args.per_gpu_train_batch_size * self.world_size
+        self.gradient_accumulation_steps = max(1, self.args.total_train_batch_size // self.train_batch_size)
+        self.total_train_batch_size = self.train_batch_size * self.gradient_accumulation_steps
+        self.steps_total = int((len(self.dataloader) // self.gradient_accumulation_steps) * self.args.num_train_epochs)
+
+        self.named_params = list(model.named_parameters())
+        self.create_optimizer_and_scheduler()
+
+        self.print_info(model, dataset_train)
```
Comment (morkovka1337) [inline misc/pytorch_toolkit/noise_suppression/train.py:264]: Please, move it to the constructor. It is strange that train function should creates something.

File: misc/pytorch_toolkit/noise_suppression/train.py [hunk 16]
```diff
@@ -0,0 +1,572 @@
...
+
+            torch.onnx.export(
+                model,
+                inputs,
+                os.path.join(args.output_dir, "model.onnx"),
+                verbose=False,
+                opset_version=11,
+                do_constant_folding=True,
+                input_names=input_names,
+                output_names=output_names)
```
Comment (morkovka1337) [inline misc/pytorch_toolkit/noise_suppression/train.py:444]: I would move export code to the extra file. Is it really needed to export model after every evaluation? Moreover, if user wants to export already trained model, they would have to retrain it again.

**Review by morkovka1337 — CHANGES_REQUESTED** [review]

File: misc/pytorch_toolkit/noise_suppression/README.md [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+4. [Convert a Model to OpenVINO™ format for Demo](#convert-a-models-to-openvino-format-for-demo)
+
+
+## Requirements
+
+To run the script Python\* 3.6 has to be installed on the machine.
+To install the required packages, run the following:
+
+```bash
+pip install -r requirements.txt
```
[the remark this answers is not in the document]
Comment (krodyush) [reply misc/pytorch_toolkit/noise_suppression/README.md:22]: it is not required to run init_venv.sh and create  virtual environment

File: misc/pytorch_toolkit/noise_suppression/evaluate.py [hunk 18]
```diff
@@ -0,0 +1,163 @@
...
+
+import models
+from metrics import sisdr
+from dataset import AudioFile
+
+logging.basicConfig(format='%(asctime)s %(levelname)s %(name)s %(message)s',
+                    datefmt='%Y-%m-%d %H:%M:%S',level=logging.INFO)
+logger = logging.getLogger('{} evaluate'.format(os.getpid()))
+def printlog(*args):
+    logger.info(' '.join([str(v) for v in args]))
```
[the remark this answers is not in the document]
Comment (krodyush) [reply misc/pytorch_toolkit/noise_suppression/evaluate.py:32]: no. the log has specific name depends on current file.

File: misc/pytorch_toolkit/noise_suppression/train.py (hunk 16 above)
Comment (krodyush) [reply misc/pytorch_toolkit/noise_suppression/train.py:444]: model is not exported after each evaluation. it is exported right after train is finished. And all trained models saved on onnx. So it is no any reason to export already trained model to onnx again.

File: misc/pytorch_toolkit/noise_suppression/evaluate.py [hunk 19]
```diff
@@ -0,0 +1,163 @@
...
+        y_clean = y_clean[delay:]
+        x_clean = x_clean[:-delay]
+        x_noisy = x_noisy[:-delay]
+
+        #calc metrics for input and output
+        file_ids.append(fi)
+        sisdr_inps_all.append(sisdr(x_noisy, x_clean).item())
+        sisdr_outs_all.append(sisdr(y_clean, x_clean).item())
+
+        sample_len = input_size/16000
```
[the remark this answers is not in the document]
Comment (krodyush) [reply misc/pytorch_toolkit/noise_suppression/evaluate.py:125]: agree. will define it somehow

File: misc/pytorch_toolkit/noise_suppression/dataset.py [hunk 20]
```diff
@@ -0,0 +1,206 @@
...
+            f_noise_idx = random.choice(self.idx_noise)
+            x = self.files_noise[f_noise_idx].read_random_segment(self.size_to_read)
+            return x
+
+        x_noise = get_noise()
+
+        nsni = self.non_stationary_noise_iter
+        # iterate to find nonstantionary noise
+        while nsni>0:
+            rms_wnd_size = int(0.050 * 16000) #50ms
```
[the remark this answers is not in the document]
Comment (krodyush) [reply misc/pytorch_toolkit/noise_suppression/dataset.py:193]: 16000 have to be defined as freq. will fix it.

File: misc/pytorch_toolkit/noise_suppression/train.py (hunk 15 above)
Comment (krodyush) [reply misc/pytorch_toolkit/noise_suppression/train.py]: done

File: misc/pytorch_toolkit/noise_suppression/dataset.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (krodyush) [reply misc/pytorch_toolkit/noise_suppression/dataset.py]: done

**Comment by krodyush** [comment]
run tests

File: misc/pytorch_toolkit/noise_suppression/train.py (hunk 16 above)
Comment (morkovka1337) [reply misc/pytorch_toolkit/noise_suppression/train.py:440]: I meant, after user trained several models, evaluated them and decided to export only the specific model. Anyway, this looks ok also.

File: misc/pytorch_toolkit/noise_suppression/train.py [hunk 21]
```diff
@@ -0,0 +1,568 @@
...
+
+        def db_to_scale(db):
+            return 10 ** (db / 20)
+
+        # scale noise to get given SNR
+        scale_noise = db_to_scale(-snr_db) * rms_clean / (rms_noise + EPS)
+
+        # do not scale noise if noise only rms_clean < -80dB
+        #noise_only_w = (rms_clean < db_to_scale(-80)).float()
+        #scale_noise = 1 * noise_only_w + scale_noise * (1 - noise_only_w)
```
Comment (morkovka1337) [inline misc/pytorch_toolkit/noise_suppression/train.py:214]: Is this needed or it can be deleted? If this is needed, please, add condition when this code works instead of uncommenting it manually.

Comment (krodyush) [reply misc/pytorch_toolkit/noise_suppression/train.py]: deleted

**Review by morkovka1337 — APPROVED** [review]
@Ilya-Krylov please review and merge
